File: android/src/main/java/expo/modules/emwdat/DisplayContentBuilder.kt
package expo.modules.emwdat

import com.meta.wearable.dat.display.views.ButtonStyle
import com.meta.wearable.dat.display.views.ContentScope
import com.meta.wearable.dat.display.views.Direction
import com.meta.wearable.dat.display.views.IconName
import com.meta.wearable.dat.display.views.IconStyle
import com.meta.wearable.dat.display.views.ImageSizePreset
import com.meta.wearable.dat.display.views.TextColor
import com.meta.wearable.dat.display.views.TextStyle

typealias InteractionHandler = (String) -> Unit

object DisplayContentBuilder {

    fun build(scope: ContentScope, node: Map<String, Any>, onInteraction: InteractionHandler) {
        buildNode(scope, node, onInteraction)
    }

    private fun buildNode(scope: ContentScope, node: Map<String, Any>, onInteraction: InteractionHandler) {
        when (node["type"] as? String) {
            "flexBox" -> buildFlexBox(scope, node, onInteraction)
            "text" -> buildText(scope, node)
            "button" -> buildButton(scope, node, onInteraction)
            "image" -> buildImage(scope, node)
            "icon" -> buildIcon(scope, node)
        }
    }

    private fun buildFlexBox(scope: ContentScope, node: Map<String, Any>, onInteraction: InteractionHandler) {
        val direction = when (node["direction"] as? String) {
            "row" -> Direction.ROW
            else -> Direction.COLUMN
        }
        val gap = (node["gap"] as? Number)?.toInt() ?: 0
        val paddingAll = (node["paddingAll"] as? Number)?.toInt() ?: 0
        val onPressId = node["onPressId"] as? String
        @Suppress("UNCHECKED_CAST")
        val children = (node["children"] as? List<*>)?.filterIsInstance<Map<String, Any>>() ?: emptyList()

        val clickHandler: (() -> Unit)? = onPressId?.let { id -> { onInteraction(id) } }

        scope.flexBox(
            direction = direction,
            gap = gap,
            paddingTop = paddingAll,
            paddingBottom = paddingAll,
            paddingStart = paddingAll,
            paddingEnd = paddingAll,
            onClick = clickHandler
        ) {
            for (child in children) {
                buildNode(this, child, onInteraction)
            }
        }
    }

    private fun buildText(scope: ContentScope, node: Map<String, Any>) {
        val content = node["content"] as? String ?: return
        val style = when (node["style"] as? String) {
            "heading" -> TextStyle.HEADING
            "meta" -> TextStyle.META
            else -> TextStyle.BODY
        }
        val color = when (node["color"] as? String) {
            "secondary" -> TextColor.SECONDARY
            else -> TextColor.PRIMARY
        }
        scope.text(content, style = style, color = color)
    }

    private fun buildButton(scope: ContentScope, node: Map<String, Any>, onInteraction: InteractionHandler) {
        val label = node["label"] as? String ?: return
        val onPressId = node["onPressId"] as? String ?: return
        val style = when (node["style"] as? String) {
            "secondary" -> ButtonStyle.SECONDARY
            "outline" -> ButtonStyle.OUTLINE
            else -> ButtonStyle.PRIMARY
        }
        scope.button(label = label, style = style, onClick = { onInteraction(onPressId) })
    }

    private fun buildImage(scope: ContentScope, node: Map<String, Any>) {
        val uri = node["uri"] as? String ?: return
        val sizePreset = when (node["sizePreset"] as? String) {
            "icon" -> ImageSizePreset.ICON
            else -> ImageSizePreset.FILL
        }
        scope.image(uri = uri, sizePreset = sizePreset)
    }

    private fun buildIcon(scope: ContentScope, node: Map<String, Any>) {
        val name = node["name"] as? String ?: return
        val iconName = parseIconName(name)
        val style = when (node["style"] as? String) {
            "outline" -> IconStyle.OUTLINE
            else -> IconStyle.FILLED
        }
        scope.icon(name = iconName, style = style)
    }

    private fun parseIconName(name: String): IconName {
        return try {
            IconName.valueOf(name.uppercase())
        } catch (e: IllegalArgumentException) {
            IconName.CHECKMARK_CIRCLE
        }
    }
}

File: ios/DisplayContentBuilder.swift
import Foundation
import MWDATDisplay

public typealias DisplayInteractionHandler = (String) -> Void

public struct DisplayContentBuilder {

    public static func build(
        from node: [String: Any],
        onInteraction: @escaping DisplayInteractionHandler
    ) -> (any ViewComponent)? {
        guard let type = node["type"] as? String else { return nil }

        switch type {
        case "flexBox":
            return buildFlexBox(node: node, onInteraction: onInteraction)
        case "text":
            return buildText(node: node)
        case "button":
            return buildButton(node: node, onInteraction: onInteraction)
        case "image":
            return buildImage(node: node)
        case "icon":
            return buildIcon(node: node)
        default:
            return nil
        }
    }

    private static func buildFlexBox(
        node: [String: Any],
        onInteraction: @escaping DisplayInteractionHandler
    ) -> FlexBox? {
        let direction: Direction = (node["direction"] as? String) == "row" ? .row : .column
        let gap = CGFloat((node["gap"] as? NSNumber)?.floatValue ?? 0)
        let paddingAll = CGFloat((node["paddingAll"] as? NSNumber)?.floatValue ?? 0)
        let onPressId = node["onPressId"] as? String
        let childNodes = node["children"] as? [[String: Any]] ?? []

        let children: [any ViewComponent] = childNodes.compactMap {
            build(from: $0, onInteraction: onInteraction)
        }

        var box = FlexBox(direction: direction, spacing: gap, padding: paddingAll > 0 ? EdgeInsets(all: paddingAll) : nil) {
            children
        }

        if let pressId = onPressId {
            box = box.onTap { onInteraction(pressId) }
        }

        return box
    }

    private static func buildText(node: [String: Any]) -> MWDATDisplay.Text? {
        guard let content = node["content"] as? String else { return nil }

        let style: TextStyle
        switch node["style"] as? String {
        case "heading": style = .heading
        case "meta": style = .meta
        default: style = .body
        }

        let color: TextColor = (node["color"] as? String) == "secondary" ? .secondary : .primary
        return MWDATDisplay.Text(content, style: style, color: color)
    }

    private static func buildButton(
        node: [String: Any],
        onInteraction: @escaping DisplayInteractionHandler
    ) -> MWDATDisplay.Button? {
        guard let label = node["label"] as? String,
              let onPressId = node["onPressId"] as? String else { return nil }

        let style: ButtonStyle
        switch node["style"] as? String {
        case "secondary": style = .secondary
        case "outline": style = .outline
        default: style = .primary
        }

        return MWDATDisplay.Button(label: label, style: style) { onInteraction(onPressId) }
    }

    private static func buildImage(node: [String: Any]) -> MWDATDisplay.Image? {
        guard let uri = node["uri"] as? String else { return nil }

        let sizePreset: ImageSizePreset = (node["sizePreset"] as? String) == "icon" ? .icon : .fill
        return MWDATDisplay.Image(uri: uri, sizePreset: sizePreset)
    }

    private static func buildIcon(node: [String: Any]) -> MWDATDisplay.Icon? {
        guard let name = node["name"] as? String else { return nil }

        let iconStyle: IconStyle = (node["style"] as? String) == "outline" ? .outline : .filled
        let iconName = parseIconName(name)
        return MWDATDisplay.Icon(name: iconName, style: iconStyle)
    }

    private static func parseIconName(_ name: String) -> IconName {
        if let match = IconName(rawValue: name) { return match }
        return .checkmarkCircle
    }
}
